lib/nodejs/sandtable.js

```diff
@@ -0,0 +1,380 @@
+  
+ 
+    var connector = (options.protocol == 'https:' ? https : http).request(options, function(serverResponse) { 
+  
+      serverResponse.pause(); 
+  
+      serverResponse.headers['access-control-allow-origin'] = '*'; 
+
+      let data = [];
+      serverResponse.on('data', chunk => {
+        data.push(chunk)
+      });
+      serverResponse.on('end', () => {
+        // save to server
+        try {
+            if ( onData != null ) {
+                onData( data );
+            }
+        } catch ( e ) {
+        }
+
+      });
+  
+      switch (serverResponse.statusCode) { 
+        // pass through.  we're not too smart here... 
+        case 200: case 201: case 202: case 203: case 204: case 205: case 206: 
+        case 304: 
+        case 400: case 401: case 402: case 403: case 404: case 405: 
+        case 406: case 407: case 408: case 409: case 410: case 411: 
+        case 412: case 413: case 414: case 415: case 416: case 417: case 418: 
+          res.writeHeader(serverResponse.statusCode, serverResponse.headers); 
+          serverResponse.pipe(res, {end:true}); 
+          serverResponse.resume(); 
+        break; 
+
+        // fix host and pass through.   
+        case 301: 
+        case 302: 
+        case 303: 
+          serverResponse.statusCode = 303; 
+          serverResponse.headers['location'] = originalHost +'/'+serverResponse.headers['location'];
+          res.writeHeader(serverResponse.statusCode);
+          serverResponse.pipe(res, {end:true}); 
+          serverResponse.resume(); 
+        break; 
+  
+ 
+        // error everything else 
+        default: 
+          var stringifiedHeaders = JSON.stringify(serverResponse.headers, null, 4); 
+          serverResponse.resume(); 
+          res.writeHeader(500);
+          res.end(process.argv.join(' ') + ':\n\nError ' + serverResponse.statusCode + '\n' + stringifiedHeaders); 
+        break;
+      } 
+      nextRequest();
+    });
+    connector.on('error', (e) => {
+      // log the error
+      console.error(`problem with request: ${e.message}`);
+      nextRequest();
+    });
+    req.pipe(connector, {end:true});
+    req.resume();
+}
+
+function parseURLParams( urlStr ) {
+    let params = {};
+    if ( urlStr != null ) {
+        const parseStr = '?params=';
+        let tokens = urlStr.split( parseStr );
+        if ( tokens[ tokens.length - 1 ] ) {
+            try {
+                params = JSON.parse( tokens[ tokens.length - 1 ] );
+            } catch ( e ) {
+                params = {};
+            }
+        }
+    }
+    return params;
+}
+
+
+// use this to save to file system
+function HandlePersistenceUpload( request, response, parsedRequest, segments ) {
+    if ( segments.length >= 2 ) {
+        var streamPath;
+        var form = new formidable.IncomingForm();
+        form.keepExtensions = true;
+        // create the path in file system if not there
+        var relPath = parsedRequest[ 'public_path' ];
```

Comment:
Where is 'public_path' defined?

Reply:
That gets set when we parse the destination url in the request.  This was already implemented and being used in other route handlers.  (see \VWF\lib\nodejs\parse-url.js)

Reply:
Thanks